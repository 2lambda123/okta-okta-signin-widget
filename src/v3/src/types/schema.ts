/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  IdxMessage,
  IdxTransaction,
  Input,
  NextStep,
  WebauthnVerificationValues,
} from '@okta/okta-auth-js';
import { IdxOption } from '@okta/okta-auth-js/lib/idx/types/idx-js';
import { FunctionComponent } from 'preact';

import { ClickHandler } from './handlers';
import { ListItem, PasswordSettings } from './password';
import { UserInfo } from './userInfo';

export type FormBag = {
  schema: Record<string, unknown>;
  uischema: UISchemaLayout;
  data: Record<string, unknown>;
  // temp schema bag to handle client validation and form submission
  dataSchema: Record<string, DataSchema>;

};

export type AutoCompleteValue = 'username'
| 'current-password'
| 'one-time-code'
| 'new-password'
| 'tel-national'
| 'given-name'
| 'family-name'
| 'email';

export type InputAttributes = { autocomplete?: AutoCompleteValue; };

// flat params
export type ActionParams = {
  [key: string]: string | boolean | number;
};

export interface ActionOptions {
  actionParams?: ActionParams;
  step: string;
  // deprecate this field once auth-js can support `step` for all scenarios
  action?: NextStep['action'];
}

/**
 * WebAuthNEnrollmentPayload
 */
export type WebAuthNEnrollmentPayload = {
  credentials: {
    /**
     * Represents the client data that was passed
     * to CredentialsContainer.create()
     */
    clientData: string;
    /**
     * BtoA String containing authenticator data and an attestation statement
     * for a newly-created key pair.
     */
    attestation: string;
  }
};

/**
 * WebAuthNVerificationPayload
 */
export type WebAuthNVerificationPayload = {
  credentials: WebauthnVerificationValues
};

export type WebAuthNEnrollmentHandler = (transaction: IdxTransaction) =>
Promise<WebAuthNEnrollmentPayload>;

export type WebAuthNAuthenticationHandler = (transaction: IdxTransaction) =>
Promise<WebAuthNVerificationPayload>;

export interface UISchemaElement {
  scope?: string; // TODO: remove
  type: string;
  label?: string;
}

export interface UISchemaLayout {
  type: UISchemaLayoutType;
  elements: (UISchemaElement | UISchemaLayout)[];
  options?: {
    onClick?: ClickHandler;
  }
}

export type PickerSchema = {
  tester: (schema: UISchemaElement) => boolean;
  mapper?: (schema: UISchemaElement) => UISchemaElement;
};
export interface CustomLayout {
  type: UISchemaLayoutType;
  elements: (CustomLayout | UISchemaElement | PickerSchema)[];
}

export enum UISchemaLayoutType {
  HORIZONTAL = 'HorizontalLayout',
  VERTICAL = 'VerticalLayout',
  STEPPER = 'Stepper',
}

export function isUISchemaLayoutType(type: string) {
  return Object.values(UISchemaLayoutType).includes(type as UISchemaLayoutType);
}

export interface FieldElement extends UISchemaElement {
  name: string;
  // TODO: refactor - dump inputMeta at top level of options
  // TODO: only use limited field as i18n field
  options: {
    inputMeta: Input;
    format?: string;
    attributes?: InputAttributes;
    defaultOption?: string | number | boolean;
    type?: string;
    customOptions?: IdxOption[],
    translations?: TranslationInfo[];
    hint?: string;
  };
}

export interface InputTextElement extends FieldElement {
  options: FieldElement['options'] & {
    attributes: InputAttributes;
  };
}
// https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#attr-type
export enum ButtonType {
  SUBMIT = 'submit',
  BUTTON = 'button',
  RESET = 'reset',
}

// TODO: use type instead of format in tester function
export interface ButtonElement extends UISchemaElement {
  type: 'Button',
  options: ActionOptions & {
    type: ButtonType;
    variant?: 'primary' | 'floating' | 'secondary';
    wide?: boolean;
    deviceChallengeUrl?: string;
    includeData?: boolean;
    dataType?: 'cancel' | 'save';
    dataSe?: string;
    stepToRender?: string;
  };
}

export interface AuthenticatorButtonElement {
  type: 'AuthenticatorButton';
  label: string;
  options: Omit<ButtonElement['options'], 'type' | 'step'> & {
    key: string;
    ctaLabel: string;
    description?: string;
  };
}

export interface WebAuthNButtonElement extends UISchemaElement {
  type: 'WebAuthNSubmitButton';
  options: {
    label: string;
    step: string;
    onClick: (() => Promise<WebAuthNEnrollmentPayload>)
    | (() => Promise<WebAuthNVerificationPayload>)
    submitOnLoad?: boolean;
    showLoadingIndicator?: boolean;
  };
}

export interface TitleElement extends UISchemaElement {
  type: 'Title';
  options: {
    content: string;
  };
}

export interface HeadingElement extends UISchemaElement {
  type: 'Heading';
  options: {
    level: 1 | 2 | 3 | 4 | 5 | 6;
    visualLevel: 1 | 2 | 3 | 4 | 5 | 6;
    content: string;
  };
}

export interface DescriptionElement extends UISchemaElement {
  type: 'Description';
  options: {
    content: string;
  };
}

export interface ReminderElement extends UISchemaElement {
  type: 'Reminder';
  options: {
    /**
     * The call to action text in the reminder content area
     */
    ctaText: string;
    /**
     * Override the default timeout before reminder appears
     */
    timeout?: number;
    linkLabel?: string;
    action?: NextStep['action'];
  };
}

export interface ListElement extends UISchemaElement {
  type: 'List';
  options: {
    items: string[];
    type?: 'unordered' | 'ordered' | 'description';
    description?: string;
  };
}

export interface PasswordRequirementsElement extends UISchemaElement {
  type: 'PasswordRequirements',
  options: {
    id: string;
    header: string;
    userInfo: UserInfo;
    settings: PasswordSettings;
    requirements: ListItem[];
    validationDelayMs: number;
  }
}

export interface LinkElement extends UISchemaElement {
  type: 'Link';
  options: ActionOptions & {
    label: string;
    href?: string;
    dataSe?: string;
  };
}

export interface ImageWithTextElement extends UISchemaElement {
  type: 'ImageWithText';
  options: {
    id: string;
    SVGIcon: FunctionComponent;
    textContent: string;
  };
}

export interface QRCodeElement extends UISchemaElement {
  type: 'QRCode';
  options: {
    label: string;
    data: string;
  };
}

export interface SpinnerElement extends UISchemaElement {
  type: 'Spinner';
  options: {
    label: string;
    valueText: string;
  };
}

export interface InfoboxElement extends UISchemaElement {
  options: {
    message: string;
    class: string;
    contentType: string;
    title?: string;
  }
}

export interface SuccessCallback extends UISchemaElement {
  options: {
    data: Record<string, unknown>;
  }
}

export type StepperNavButtonConfigDirection = 'next' | 'prev';

export type StepperNavButtonConfigAttrs = {
  variant?: 'primary' | 'secondary';
  label?: string;
  id?: string;
};

export interface StepperLayout {
  type: UISchemaLayoutType.STEPPER;
  elements: UISchemaLayout[];
  options?: {
    defaultStepIndex: () => number;
  }
}

export interface StepperButtonElement {
  type: 'StepperButton',
  label: string;
  options: Omit<ButtonElement['options'], 'step'>
  & {
    nextStepIndex: number;
  }
}

export interface StepperRadioElement {
  type: 'StepperRadio',
  options: {
    customOptions: Array<IdxOption & { key?: string; }>,
    name: string;
    defaultOption: string | number | boolean;
  }
}

export interface RedirectElement extends UISchemaElement {
  type: 'Redirect',
  options: { url: string; },
}

type ValidateFunction = (data: FormBag['data']) => Partial<IdxMessage>[]
| undefined;

export interface DataSchema {
  validate?: ValidateFunction;
}

export interface TranslationInfo {
  name: string;
  i18nKey: string;
  value: string;
}

export interface PasswordWithConfirmationElement extends UISchemaElement {
  type: 'PasswordWithConfirmation';
  options: {
    input: FieldElement,
  };
}
