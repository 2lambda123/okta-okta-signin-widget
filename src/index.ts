// Entry point for Typescript build
import OktaSignIn from './widget/OktaSignIn';

export default OktaSignIn;
export { OktaSignIn };

export * from './types';
