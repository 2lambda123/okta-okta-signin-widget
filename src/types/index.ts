export * from './api';
export * from './errors';
export * from './registration';
export * from './options';
export * from './results';
