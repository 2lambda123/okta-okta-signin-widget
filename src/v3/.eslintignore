!.*.json
!.eslintrc.js
!tsconfig.json
assets/
build/
build2/
coverage/
dist/
mocks/
node_modules/
packages/**/*.js
scripts/
size-plugin.json
src/properties/json/
src/static/mockServiceWorker.js
static/
target
test-reports/
test/unit/helpers/xhr/
vendor/
