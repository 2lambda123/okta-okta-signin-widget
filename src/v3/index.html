<html lang="en">

<head>
  <title>SIW G3 Playground</title>
  <meta charset="UTF-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <!-- FIXME /css/okta-sign-in.next.css -->
  <link href="/style.css" type="text/css" rel="stylesheet" />
  <!-- <script src="/js/okta-sign-in.next.min.js"></script> -->
  <script type="module" src="./src/index.ts"></script>
</head>

<body>
  <div id="okta-login-container"></div>
  <script type="module" src="./src/playground.ts"></script>
</body>

</html>
