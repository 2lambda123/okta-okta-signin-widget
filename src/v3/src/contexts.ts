/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxMessage, IdxTransaction, OktaAuth } from '@okta/okta-auth-js';
import { createContext } from 'preact';
import { MutableRef, StateUpdater, useContext } from 'preact/hooks';

import {
  ActionOptions,
  FormBag,
  WidgetProps,
} from './types';

// Widget context
type IWidgetContext = {
  authClient: OktaAuth;
  widgetProps: WidgetProps;
  setMessages: StateUpdater<IdxMessage[]>;
  // // TODO: OKTA-502849 - Update param type
  // (RenderSuccessCallback / RenderErrorCallback) once merged into okta-signin-widget
  onSuccessCallback?: (data: Record<string, unknown>) => void;
  onErrorCallback?: (data: Record<string, unknown>) => void;
  idxTransaction: IdxTransaction | undefined;
  setIdxTransaction: StateUpdater<IdxTransaction | undefined>;
  stepToRender: string | undefined;
  setStepToRender: StateUpdater<string | undefined>;
  data: Record<string, unknown>;
  setData: StateUpdater<Record<string, unknown>>;
  dataSchemaRef: MutableRef<FormBag['dataSchema'] | undefined>;
};

const createWidgetContext = <T extends unknown>() => {
  // Create a context with a generic parameter or undefined
  const genericContext = createContext<T | undefined>(undefined);

  // Check if the value provided to the context is defined or throw an error
  const useWidgetContext = () => {
    const contextIsDefined = useContext(genericContext);
    if (!contextIsDefined) {
      throw new Error('useWidgetContext must be used within a Provider');
    }
    return contextIsDefined;
  };

  return [useWidgetContext, genericContext.Provider] as const;
};

const [useWidgetContext, WidgetContextProvider] = createWidgetContext<IWidgetContext>();

export { useWidgetContext, WidgetContextProvider };

// Form context
type IFormContext = {
  submissionOptionsRef: MutableRef<ActionOptions | undefined>;
};

export const FormContext = createContext<IFormContext>({} as IFormContext);

export const useFormContext = () => useContext(FormContext);

// Stepper context
type IStepperContext = {
  setStepIndex: StateUpdater<number | undefined>;
};

export const StepperContext = createContext<IStepperContext>({
  setStepIndex: () => {},
});

export const useStepperContext = () => useContext(StepperContext);
