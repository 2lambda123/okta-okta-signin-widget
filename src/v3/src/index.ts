/*
 * Copyright (c) 2022-present, Okta, Inc. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant
 * to the Apache License, Version 2.0 (the "License.")
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { config } from '../widgetrc';
import { OktaSignIn } from './OktaSignIn';

// if (process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'test') {
//   import('./mocks/browser')
//     .then(({ getWorker }) => getWorker())
//     .then((worker) => worker?.start());
// }

if (import.meta.env.MODE === 'development') {
  // if (import.meta.env.VITE_MSW) {
  //   const { worker } = await import('../mocks/browser');
  //   worker.start({ quiet: import.meta.env.VITE_MSW_QUIET });
  // }
  
  // eslint-disable-next-line no-new
  new OktaSignIn(config);
}
