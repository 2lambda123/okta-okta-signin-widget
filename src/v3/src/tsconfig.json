{
  "extends": "../tsconfig.base.json",
  "include": [
    "./**/PhoneAuthenticator.test.tsx"
  ]
}
