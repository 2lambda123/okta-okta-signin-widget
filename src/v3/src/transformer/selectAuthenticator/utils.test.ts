/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { Input } from '@okta/okta-auth-js';
import { IdxOption } from '@okta/okta-auth-js/lib/idx/types/idx-js';
import { AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP, AUTHENTICATOR_KEY } from 'src/constants';

import {
  getAuthenticatorEnrollButtonElements,
  getAuthenticatorVerifyButtonElements,
  getOVMethodTypeAuthenticatorButtonElements,
  isOnlyPushWithAutoChallenge,
} from './utils';

describe('Select Authenticator Utility Tests', () => {
  describe('getOVMethodTypeAuthenticatorButtonElements Tests', () => {
    it('should return an empty array when an empty array of options is provided', () => {
      expect(getOVMethodTypeAuthenticatorButtonElements([])).toEqual([]);
    });

    it('should return formatted Authenticator Option Values '
      + 'when OV methodTypes are provided', () => {
      const options = [
        { label: 'Enter a code', value: 'totp' } as IdxOption,
        { label: 'Get a push notification', value: 'push' } as IdxOption,
      ];
      expect(getOVMethodTypeAuthenticatorButtonElements(options)).toEqual([
        {
          type: 'AuthenticatorButton',
          label: options[0].label,
          options: {
            key: AUTHENTICATOR_KEY.OV,
            ctaLabel: 'oie.verify.authenticator.button.text',
            idxMethodParams: {
              authenticator: {
                methodType: options[0].value,
              },
            },
          },
        },
        {
          type: 'AuthenticatorButton',
          label: options[1].label,
          options: {
            key: AUTHENTICATOR_KEY.OV,
            ctaLabel: 'oie.verify.authenticator.button.text',
            idxMethodParams: {
              authenticator: {
                methodType: options[1].value,
              },
            },
          },
        },
      ]);
    });
  });

  describe('isOnlyPushWithAutoChallenge Tests', () => {
    it('should return false when authenticator inputs are not provided', () => {
      expect(isOnlyPushWithAutoChallenge(undefined)).toBe(false);
    });

    it('should return false when authenticator input values do not contain methodType', () => {
      const authenticatorInputs: Input[] = [{ name: 'autoChallenge' }];
      expect(isOnlyPushWithAutoChallenge(authenticatorInputs)).toBe(false);
    });

    it('should return false when authenticator input values do not contain autoChallenge', () => {
      const authenticatorInputs: Input[] = [{ name: 'methodType' }];
      expect(isOnlyPushWithAutoChallenge(authenticatorInputs)).toBe(false);
    });

    it('should return false when authenticator input values contains multiple methodType options', () => {
      const authenticatorInputs: Input[] = [
        {
          name: 'methodType',
          options: [{ label: 'Code', value: 'totp' }, { label: 'Push', value: 'push' }],
        },
        { name: 'autoChallenge' },
      ];
      expect(isOnlyPushWithAutoChallenge(authenticatorInputs)).toBe(false);
    });

    it('should return false when authenticator input values contain only totp methodType', () => {
      const authenticatorInputs: Input[] = [
        {
          name: 'methodType',
          options: [{ label: 'Code', value: 'totp' }],
        },
        { name: 'autoChallenge' },
      ];
      expect(isOnlyPushWithAutoChallenge(authenticatorInputs)).toBe(false);
    });

    it('should return true when authenticator input values contain only "push" methodType '
      + 'and has autoChallenge input', () => {
      const authenticatorInputs: Input[] = [
        {
          name: 'methodType',
          options: [{ label: 'Push', value: 'push' }],
        },
        { name: 'autoChallenge' },
      ];
      expect(isOnlyPushWithAutoChallenge(authenticatorInputs)).toBe(true);
    });
  });

  describe('getAuthenticatorVerifyButtonElements Tests', () => {
    it('should return empty array when no authenticator options are provided', () => {
      expect(getAuthenticatorVerifyButtonElements([])).toEqual([]);
    });

    it('should return formatted authenticator options when raw options are provided', () => {
      const mockPhoneNumber = '2XXXXXX123';
      const options: IdxOption[] = Object.entries(AUTHENTICATOR_KEY)
        .filter(([, value]) => value !== AUTHENTICATOR_KEY.OV
          && value !== AUTHENTICATOR_KEY.DEFAULT)
        .map(([key]) => {
          const option = {
            label: key,
            value: [{ name: 'methodType', value: AUTHENTICATOR_KEY[key] }],
            relatesTo: {
              id: '',
              type: '',
              methods: [{ type: '' }],
              displayName: '',
              key: AUTHENTICATOR_KEY[key],
              profile: {},
            },
          };
          if (key === AUTHENTICATOR_KEY.PHONE) {
            option.relatesTo.profile = { phoneNumber: mockPhoneNumber };
          }
          return option;
        });
      const authenticatorOptionValues = getAuthenticatorVerifyButtonElements(options);

      expect(authenticatorOptionValues.length).toBe(15);
      options.forEach((option) => {
        const currentOption = authenticatorOptionValues
          .find(({ options: { key: authKey } }) => authKey === option.relatesTo?.key);
        expect(currentOption?.options.key).toBe(option.relatesTo?.key);
        expect(currentOption?.label).toBe(option.label);
        expect(currentOption?.options.ctaLabel)
          .toBe('oie.verify.authenticator.button.text');
        expect(currentOption?.options.description)
          .toBe(option.value === AUTHENTICATOR_KEY.PHONE ? mockPhoneNumber : undefined);
        expect(currentOption?.options.descriptionParams).toBeUndefined();
      });
    });

    it('should return formatted Okta Verify method type options '
      + 'when raw options are provided', () => {
      const options: IdxOption[] = [{
        label: 'Okta Verify',
        value: [
          {
            name: 'methodType',
            options: [
              { label: 'Code', value: 'totp' },
              { label: 'Push', value: 'push' },
            ],
          },
          { name: 'id', value: '1234abc' },
        ],
        relatesTo: {
          id: '',
          type: '',
          methods: [{ type: '' }],
          displayName: '',
          key: AUTHENTICATOR_KEY.OV,
        },
      }];

      const authenticatorOptionValues = getAuthenticatorVerifyButtonElements(options);

      expect(authenticatorOptionValues.length).toBe(2);
      expect(authenticatorOptionValues[0].options.key).toBe(AUTHENTICATOR_KEY.OV);
      expect(authenticatorOptionValues[0].label).toBe('Code');
      expect(authenticatorOptionValues[0].options.ctaLabel).toBe('oie.verify.authenticator.button.text');
      expect(authenticatorOptionValues[0].options.idxMethodParams.authenticator.methodType).toBe('totp');
      expect(authenticatorOptionValues[1].options.key).toBe(AUTHENTICATOR_KEY.OV);
      expect(authenticatorOptionValues[1].label).toBe('Push');
      expect(authenticatorOptionValues[1].options.ctaLabel).toBe('oie.verify.authenticator.button.text');
      expect(authenticatorOptionValues[1].options.idxMethodParams.authenticator.methodType).toBe('push');
    });
  });

  describe('getAuthenticatorEnrollButtonElements Tests', () => {
    it('should return empty array when no authenticator options are provided', () => {
      expect(getAuthenticatorEnrollButtonElements([])).toEqual([]);
    });

    it('should return formatted authenticator options when raw options are provided', () => {
      const mockPhoneNumber = '2XXXXXX123';
      const options: IdxOption[] = Object.entries(AUTHENTICATOR_KEY)
        .filter(([, value]) => value !== AUTHENTICATOR_KEY.OV
          && value !== AUTHENTICATOR_KEY.DEFAULT)
        .map(([key]) => {
          const option = {
            label: key,
            value: [{ name: 'methodType', value: AUTHENTICATOR_KEY[key] }],
            relatesTo: {
              id: '',
              type: '',
              methods: [{ type: '' }],
              displayName: '',
              key: AUTHENTICATOR_KEY[key],
              profile: {},
            },
          };
          if (key === AUTHENTICATOR_KEY.PHONE) {
            option.relatesTo.profile = { phoneNumber: mockPhoneNumber };
          }
          return option;
        });
      const authenticatorOptionValues = getAuthenticatorEnrollButtonElements(options);

      expect(authenticatorOptionValues.length).toBe(15);
      options.forEach((option) => {
        const currentOption = authenticatorOptionValues
          .find(({ options: { key: authKey } }) => authKey === option.relatesTo?.key);
        expect(currentOption?.options.key).toBe(option.relatesTo?.key);
        expect(currentOption?.label).toBe(option.label);
        expect(currentOption?.options.ctaLabel)
          .toBe('oie.enroll.authenticator.button.text');
        expect(currentOption?.options.description)
          .toBe(option.relatesTo?.key === AUTHENTICATOR_KEY.ON_PREM
            ? 'next.oie.on_prem.authenticator.default.description'
            : AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP[option.relatesTo?.key as string]);
        expect(currentOption?.options.descriptionParams).toBeUndefined();
      });
    });

    it('should format on_prem authenticator option when display name exists', () => {
      const displayName = 'Some Company Name';
      const options: IdxOption[] = [{
        label: 'On Prem',
        value: [{ name: 'methodType', value: AUTHENTICATOR_KEY.ON_PREM }],
        relatesTo: {
          id: '',
          type: '',
          methods: [{ type: '' }],
          displayName,
          key: AUTHENTICATOR_KEY.ON_PREM,
        },
      }];
      const authenticatorOptionValues = getAuthenticatorEnrollButtonElements(options);

      expect(authenticatorOptionValues.length).toBe(1);
      expect(authenticatorOptionValues[0].options.key).toBe(AUTHENTICATOR_KEY.ON_PREM);
      expect(authenticatorOptionValues[0].label).toBe('On Prem');
      expect(authenticatorOptionValues[0].options.ctaLabel)
        .toBe('oie.enroll.authenticator.button.text');
      expect(authenticatorOptionValues[0].options.description)
        .toBe(AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP[AUTHENTICATOR_KEY.ON_PREM]);
      expect(authenticatorOptionValues[0].options.descriptionParams).toEqual([displayName]);
    });

    it('should return formatted Okta Verify method type options '
      + 'when raw options are provided', () => {
      const options: IdxOption[] = [{
        label: 'Okta Verify',
        value: [
          {
            name: 'methodType',
            options: [
              { label: 'Code', value: 'totp' },
              { label: 'Push', value: 'push' },
            ],
          },
          { name: 'id', value: '1234abc' },
        ],
        relatesTo: {
          id: '',
          type: '',
          methods: [{ type: '' }],
          displayName: '',
          key: AUTHENTICATOR_KEY.OV,
        },
      }];

      const authenticatorOptionValues = getAuthenticatorEnrollButtonElements(options);

      expect(authenticatorOptionValues.length).toBe(2);
      expect(authenticatorOptionValues[0].options.key).toBe(AUTHENTICATOR_KEY.OV);
      expect(authenticatorOptionValues[0].label).toBe('Code');
      expect(authenticatorOptionValues[0].options.ctaLabel).toBe('oie.enroll.authenticator.button.text');
      expect(authenticatorOptionValues[0].options.idxMethodParams.authenticator.methodType).toBe('totp');
      expect(authenticatorOptionValues[1].label).toBe('Push');
      expect(authenticatorOptionValues[1].options.ctaLabel).toBe('oie.enroll.authenticator.button.text');
      expect(authenticatorOptionValues[1].options.idxMethodParams.authenticator.methodType).toBe('push');
    });
  });
});
