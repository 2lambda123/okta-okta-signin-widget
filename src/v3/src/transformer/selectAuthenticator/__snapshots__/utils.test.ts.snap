// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Authenticator Utility Tests getAuthenticatorEnrollButtonElements Tests should format on_prem authenticator option when display name exists 1`] = `
Array [
  Object {
    "label": "On Prem",
    "options": Object {
      "actionParams": Object {
        "authenticator.enrollmentId": undefined,
        "authenticator.id": undefined,
        "authenticator.methodType": "onprem_mfa",
      },
      "authenticator": Object {
        "displayName": "Some Company Name",
        "id": "",
        "key": "onprem_mfa",
        "methods": Array [
          Object {
            "type": "",
          },
        ],
        "type": "",
      },
      "ctaLabel": "oie.enroll.authenticator.button.text",
      "dataSe": "onprem_mfa-onprem_mfa",
      "description": "oie.on_prem.authenticator.description",
      "iconName": "onprem_mfa_0",
      "includeData": true,
      "includeImmutableData": false,
      "key": "onprem_mfa",
      "logoUri": undefined,
      "step": "select-authenticator-enroll",
      "type": "button",
      "usageDescription": undefined,
    },
    "type": "AuthenticatorButton",
  },
]
`;

exports[`Select Authenticator Utility Tests getAuthenticatorEnrollButtonElements Tests should return formatted Okta Verify method type options when raw options are provided 1`] = `
Array [
  Object {
    "label": "Code",
    "options": Object {
      "actionParams": Object {
        "authenticator.id": "1234abc",
        "authenticator.methodType": "totp",
      },
      "ctaLabel": "oie.enroll.authenticator.button.text",
      "dataSe": "okta_verify-totp",
      "description": "oie.okta_verify.authenticator.description",
      "iconDescr": "factor.totpSoft.description",
      "iconName": "oktaVerify",
      "includeData": true,
      "includeImmutableData": false,
      "key": "okta_verify",
      "step": "select-authenticator-enroll",
      "type": "button",
    },
    "type": "AuthenticatorButton",
  },
  Object {
    "label": "Push",
    "options": Object {
      "actionParams": Object {
        "authenticator.id": "1234abc",
        "authenticator.methodType": "push",
      },
      "ctaLabel": "oie.enroll.authenticator.button.text",
      "dataSe": "okta_verify-push",
      "description": "oie.okta_verify.authenticator.description",
      "iconDescr": "factor.push.description",
      "iconName": "oktaVerifyPush",
      "includeData": true,
      "includeImmutableData": false,
      "key": "okta_verify",
      "step": "select-authenticator-enroll",
      "type": "button",
    },
    "type": "AuthenticatorButton",
  },
]
`;

exports[`Select Authenticator Utility Tests getAuthenticatorVerifyButtonElements Tests should return formatted Okta Verify method type options when raw options are provided 1`] = `
Array [
  Object {
    "label": "Code",
    "options": Object {
      "actionParams": Object {
        "authenticator.id": "1234abc",
        "authenticator.methodType": "totp",
      },
      "ctaLabel": "oie.verify.authenticator.button.text",
      "dataSe": "okta_verify-totp",
      "description": "oie.okta_verify.label",
      "iconDescr": "factor.totpSoft.description",
      "iconName": "oktaVerify",
      "includeData": true,
      "includeImmutableData": false,
      "key": "okta_verify",
      "step": "select-authenticator-enroll",
      "type": "button",
    },
    "type": "AuthenticatorButton",
  },
  Object {
    "label": "Push",
    "options": Object {
      "actionParams": Object {
        "authenticator.id": "1234abc",
        "authenticator.methodType": "push",
      },
      "ctaLabel": "oie.verify.authenticator.button.text",
      "dataSe": "okta_verify-push",
      "description": "oie.okta_verify.label",
      "iconDescr": "factor.push.description",
      "iconName": "oktaVerifyPush",
      "includeData": true,
      "includeImmutableData": false,
      "key": "okta_verify",
      "step": "select-authenticator-enroll",
      "type": "button",
    },
    "type": "AuthenticatorButton",
  },
]
`;
